swarm.1
.\" ============================================================================
.TH swarm 1 "February 26, 2015" "version 2.0.1" "USER COMMANDS"
.\" ============================================================================
.SH NAME
swarm \(em find clusters of nearly-identical nucleotide amplicons
.\" ============================================================================
.SH SYNOPSIS
.B swarm
[
.I options
] [
.I filename
]
.\" ============================================================================
.SH DESCRIPTION
Environmental or clinical molecular studies generate large volumes of
amplicons (e.g. SSU-rRNA sequences) that need to be clustered into
molecular operational taxonomic units. Traditional clustering methods
are based on greedy, input-order dependent algorithms, with arbitrary
selection of global cluster size and cluster centroids. To address
that problem, we developed \fBswarm\fR, a fast and robust method that
recursively groups amplicons with \fId\fR or less
differences. \fBswarm\fR produces stable clusters (or "swarms"), free
from centroid selection induced input-order dependency.
.PP
Exact clustering is impractical on large data sets when using a naïve
all-vs-all approach (more precisely a 2-combination without
repetitions), as it implies unrealistic numbers of pairwise
comparisons. \fBswarm\fR is based on a maximum number of differences
\fId\fR, and focuses only on very close relationships. An efficient
\fIk\fR-mer-based filtering and an astute use of comparisons results
obtained during the process allows to avoid most of the amplicon
comparisons needed in a naïve approach. To speed up the remaining
amplicon comparisons, \fBswarm\fR implements an extremely fast
Needleman-Wunsch algorithm making use of the Streaming SIMD Extensions
(SSE2) of modern x86-64 CPUs. If SSE2 instructions are not available,
\fBswarm\fR exits with an error message.
.PP
By default, \fBswarm\fR will work at its finest possible resolution
(\fId\fR = 1). It will then use a specifically crafted "stepping
stone" algorithm that provides indentical results, but at a much
greater speed. The speed difference is due to a change in
computational complexity (quadratic vs. linear), and will increase
favorably for \fBswarm\fR as datasets will grow larger.
.PP
\fBswarm\fR reads the named input \fIfilename\fR, a fasta file of
nucleotide amplicons. The amplicon identifier is defined as the string
comprised between the ">" symbol and the first space or the end of the
line, whichever comes first. As \fBswarm\fR outputs lists of amplicon
identifiers, amplicon identifiers must be unique to avoid ambiguity;
swarm exits with an error message if identifiers are not unique. If
amplicon identifiers end with a "_" followed by a positive integer,
that value is used as the amplicon copy number in the statistics
output file (see the option \-z to change that behavior for
usearch/vsearch users). The amplicon sequence is defined as a string
of [acgt] or [acgu] symbols (case insensitive), starting after the end
of the identifier line and ending before the next identifier line or
the file end; \fBswarm\fR exits with an error message if any other
symbol is present. Default is to read from standard input if no file
is named, or if the file name is "\-".
.\" ----------------------------------------------------------------------------
.SS Options
\fBswarm\fR recognizes the following command-line options:
.RS
.TP 9
.BI \-b\fP,\fB\ \-\-boundary\~ "positive integer"
when using the option \-\-fastidious, define what is a small OTU. By
default, an OTU is small if it has a mass of 2 or less, meaning that
it is composed of either one amplicon of abundance 2, or two amplicons
of abundance 1. Any positive value greater than 2 can be used. Using
higher boundary values will speed up the second pass, but also reduce
the taxonomical resolution of swarm results.
.TP
.BI \-d\fP,\fB\ \-\-differences\~ "positive integer"
maximum number of differences allowed between two amplicons, meaning
that two amplicons will be grouped if they have \fIinteger\fR (or
less) differences. This is \fBswarm\fR's most important parameter. The
number of differences is calculated as the number of mismatches
(substitutions, insertions or deletions) between the two amplicons
once the optimal pairwise global alignment has been found (see
"advanced options" for parameters influencing the pairwise
alignment). Any \fIinteger\fR between 1 and 256 can be used, but
aligning correctly two very distant amplicons is difficult and results
should be considered with caution. Using high \fId\fR values will also
decrease the taxonomical resolution of swarm results. Default number
of differences is 1.
.TP
.BI \-f\fP,\fB\ \-\-fastidious \0filename
when working with \fId\fR = 1, perform a second pass to reduce the
number of small OTUs. During the stone stepping process at \fId\fR =
1, a step can be missing for purely stochastic reasons, interrupting
the agglomerative process. That option will create virtual steps to
graft small OTUs upon bigger ones. By default, an OTU is small if it
has a mass of 2 or less (see the \-\-boundary option to increase that
value). Output the refined OTUs to \fIfilename\fR. Warning, that
second pass can be many times slower than the first pass.
.TP
.B \-h\fP,\fB\ \-\-help
display this help and exit.
.TP
.BI \-i\fP,\fB\ \-\-internal\-structure \0filename
output all pairs of nearly-identical amplicons to \fIfilename\fR using
a five-columns tab-delimited format:
.RS
.RS
.nr step 1 1
.IP \n[step]. 4
amplicon A label.
.IP \n+[step].
amplicon B label.
.IP \n+[step].
number of differences between amplicons A and B (\fIpositive
integer\fR).
.IP \n+[step].
OTU number (\fIpositive integer\fR). OTUs are numbered in their order
of delineation, starting from 1. All pairs of amplicons belonging to
the same OTU will receive the same number.
.IP \n+[step].
number of steps from the OTU seed to amplicon B (\fIpositive
integer\fR).
.RE
.RE
.TP
.BI \-l\fP,\fB\ \-\-log \0filename
output all messages to \fIfilename\fR instead of \fIstandard error\fR,
with the exception of error messages of course. That option is useful
in situations where writing to \fIstandard error\fR is problematic
(for example, with certain job schedulers).
.TP
.B \-n\fP,\fB\ \-\-no\-otu\-breaking
desactivate the built-in OTU refinement (not recommended). Amplicon
abundance values are used to identify transitions among in-contact
OTUs and to separate them, yielding higher-resolution clustering
results. That option prevents that, and in practice, allows the
creation of a link between amplicon A and B, even if the abundance of
B is higher than the abundance of A.
.TP
.BI \-o\fP,\fB\ \-\-output\-file \0filename
output result to \fIfilename\fR. Result is a list of swarms, one swarm
per line. A swarm is a list of amplicon identifiers separated by
spaces. Default is to write to standard output.
.TP
.B \-r\fP,\fB\ \-\-mothur
output results in a format compatible with Mothur. That option
modifies swarm's default output format.
.TP
.BI \-s\fP,\fB\ \-\-statistics\-file \0filename
output statistics to the specified file. Default is not to output
statistics.  The file is a tab-separated table with one swarm per row
and seven columns of information: number of unique amplicons in the
swarm, total copy number of amplicons in the swarm, identifier of the
initial seed, initial seed copy number (if applicable), number of
singletons (amplicons with a copy number of 1), maximum number of
generations (i.e. numbers of iterations before the swarm reached its
natural limits), and the maximum radius of the swarm (i.e. number of
differences between the seed and the furthermost amplicon in the
swarm).
.TP
.BI \-t\fP,\fB\ \-\-threads\~ "positive integer"
number of computation threads to use. The number of threads should be
lesser or equal to the number of available CPU cores. Default number
of threads is 1.
.TP
.BI \-u\fP,\fB\ \-\-uclust\-file \0filename
output results in uclust-like file format to the specified file. That
option does not modify swarm default output format.
.TP
.B \-v\fP,\fB\ \-\-version
output version information and exit.
.TP
.B \-z\fP,\fB\ \-\-usearch\-abundance
accept amplicon abundances specified using the usearch/vsearch style
(">label;size=INT[;]").
.LP
.\" ----------------------------------------------------------------------------
.SS Advanced options
\fBswarm\fR recognizes advanced command-line options modifying the
pairwise global alignment scoring parameters:
.RS
.TP 9
.BI \-m\fP,\fB\ \-\-match\-reward\~ "positive integer"
reward for a nucleotide match. Default is 5.
.TP
.BI \-p\fP,\fB\ \-\-mismatch\-penalty\~ "positive integer"
penalty for a nucleotide mismatch. Default is 4.
.TP
.BI \-g\fP,\fB\ \-\-gap\-opening\-penalty\~ "positive integer"
gap open penalty. Default is 12.
.TP
.BI \-e\fP,\fB\ \-\-gap\-extension\-penalty\~ "positive integer"
gap extension penalty. Default is 4.
.LP
As \fBswarm\fR focuses on close relationships, final results are
resilient to model parameters modifications. Modifying model
parameters only impacts analysis using a high number of differences.
.\" classic parameters are +5/-4/-12/-1
.\" ============================================================================
.SH EXAMPLES
.B swarm
-t 4 \-o
.I myfile.swarms myfile.fasta
.br
Divide the data set \fImyfile.fasta\fR into OTUs with the finest
resolution possible (1 difference, built-in breaking) using 4
computation threads and the fast algorithm. OTUs are written to the
file \fImyfile.swarms\fR.
.PP
zcat file.fas.gz | \fBswarm\fR | awk "{print NF}" | sort \-n | uniq
\-c
.br
Use swarm in a pipeline to read a compressed fasta file and to get its
swarm size profile (with default parameters).
.\" ============================================================================
.\" .SH LIMITATIONS
.\" List known limitations or bugs.
.\" ============================================================================
.SH AUTHORS
Concept by Frédéric Mahé, implementation by Torbjørn Rognes.
.\" ============================================================================
.SH CITATION
Mahé F, Rognes T, Quince C, de Vargas C, Dunthorn M. (2014) Swarm:
robust and fast clustering method for amplicon-based
studies. \fIPeerJ\fR 2:e593 <http://dx.doi.org/10.7717/peerj.593>
.\" ============================================================================
.SH REPORTING BUGS
Submit suggestions and bug-reports at
<https://github.com/torognes/swarm/issues>, send a pull request on
<https://github.com/torognes/swarm>, or compose a friendly or
curmudgeonly e-mail to Frédéric Mahé <mahe@rhrk.uni-kl.de> and
Torbjørn Rognes <torognes@ifi.uio.no>.
.\" ============================================================================
.SH AVAILABILITY
The software is available from <https://github.com/torognes/swarm>
.\" ============================================================================
.SH COPYRIGHT
Copyright (C) 2012, 2013, 2014 Frédéric Mahé & Torbjørn Rognes
.PP
This program is free software: you can redistribute it and/or modify
it under the terms of the GNU Affero General Public License as
published by the Free Software Foundation, either version 3 of the
License, or any later version.
.PP
This program is distributed in the hope that it will be useful, but
WITHOUT ANY WARRANTY; without even the implied warranty of
MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
Affero General Public License for more details.
.PP
You should have received a copy of the GNU Affero General Public
License along with this program.  If not, see
<http://www.gnu.org/licenses/>.
.PP
.\" ============================================================================
.SH SEE ALSO
\fBswipe\fR, an extremely fast Smith-Waterman database search tool by
Torbjørn Rognes (available from <https://github.com/torognes/swipe>).
.PP
\fBvsearch\fR, an open-source re-implementation of the classic uclust
clustering method (by Robert C. Edgar), along with other amplicon
filtering and searching tools. \fBvsearch\fR is implemented by
Torbjørn Rognes and documented by Frédéric Mahé, and is available at
<https://github.com/torognes/vsearch>.
.PP
.\" ============================================================================
.SH VERSION HISTORY
New features and important modifications of \fBswarm\fR (short lived
or minor bug releases are not mentioned):
.RS
.TP
.BR v2.0.1\~ "released February 26, 2015"
Version 2.0.1 is a development version that contains a partial
implementation of the fastidious option, but it is not usable yet.
.TP
.BR v2.0.0\~ "released December 3, 2014"
Version 2.0.0 is faster and easier to use, providing new output
options (\-\-internal\-structure and \-\-log), new control options
(\-\-boundary, \-\-fastidious, \-\-no\-otu\-breaking), and built-in
OTU refinement. When using default parameters, a novel and
considerably faster algorithmic approach is used, guaranteeing swarm's
scalability.
.TP
.BR v1.2.20\~ "released November 6, 2014"
Version 1.2.20 presents a production-ready version of the alternative
algorithm (option \-a), with optional built-in OTU breaking (option
\-n). That alternative algorithmic approach (usable only with d = 1)
is considerably faster than currently used clustering algorithms, and
can deal with datasets of 100 million unique amplicons or more in a
few hours. Of course, results are rigourously identical to the results
previously produced with swarm. That release also introduces new
options to control swarm output (options \-i and \-l).
.TP
.BR v1.2.19\~ "released October 3, 2014"
Version 1.2.19 fixes a problem related to abundance information when
the sequence identifier includes multiple underscore characters.
.TP
.BR v1.2.18\~ "released September 29, 2014"
Version 1.2.18 reenables the possibility of reading sequences from
stdin if no file name is specified on the command line. It also fixes
a bug related to cpu features detection.
.TP
.BR v1.2.17\~ "released September 28, 2014"
Version 1.2.17 fixes a memory allocation bug introduced in version
1.2.15.
.TP
.BR v1.2.16\~ "released September 27, 2014"
Version 1.2.16 fixes a bug in the abundance sort introduced in version
1.2.15.
.TP
.BR v1.2.15\~ "released September 27, 2014"
Version 1.2.15 sorts the input sequences in order of decreasing
abundance unless they are detected to be sorted already. When using
the alternative algorithm for d = 1 it also sorts all subseeds in
order of decreasing abundance.
.TP
.BR v1.2.14\~ "released September 27, 2014"
Version 1.2.14 fixes a bug in the output with the swarm_breaker option
(\-b) when using the alternative algorithm (\-a).
.TP
.BR v1.2.12\~ "released August 18, 2014"
Version 1.2.12 introduces an option \-\-alternative\-algorithm to use
an extremely fast, experimental clustering algorithm for the special
case d = 1. Multithreading scalability of the default algorithm has
been noticeably improved.
.TP
.BR v1.2.10\~ "released August 8, 2014"
allows amplicon abundances to be specified using the usearch style in
the sequence header (e.g. ">id;size=1") when the \-z option is chosen.
.TP
.BR v1.2.8\~ "released August 5, 2014"
swarm 1.2.8 fixes an error with the gap extension penalty. Previous
versions used a gap penalty twice as large as intended. That bug
correction induces small changes in clustering results.
.TP
.BR v1.2.6\~ "released May 23, 2014"
Version 1.2.6 introduces an option \-\-mothur to output swarm results in
a format compatible with the microbial ecology community analysis
software suite Mothur.
.TP
.BR v1.2.5\~ "released April 11, 2014"
Version 1.2.5 removes the need for a POPCNT hardware instruction to be
present. Swarm now automatically checks whether POPCNT is available
and uses a slightly slower software implementation if not. Only basic
SSE2 instructions are now required to run swarm.
.TP
.BR v1.2.4\~ "released January 30, 2014"
Version 1.2.4 introduces an option \-\-break\-swarms to output all
pairs of amplicons with \fId\fR differences to standard error. That
option is used by the companion script `swarm_breaker.py` to refine
swarm results. The syntax of the inline assembly code is changed for
compatibility with more compilers.
.TP
.BR v1.2\~ "released May 16, 2013"
Version 1.2 greatly improves speed by using alignment-free comparisons
of amplicons based on \fIk\fR-mer word content. For each amplicon, the
presence-absence of all possible 5-mers is computed and recorded in a
1024-bits vector. Vector comparisons are extremely fast and
drastically reduce the number of costly pairwise alignments performed
by swarm. While remaining exact, swarm 1.2 can be more than 100-times
faster than swarm 1.1, when using a single thread with a large set of
sequences. The minor version 1.1.1, published just before, adds
compatibility with Apple computers, and corrects an issue in the
pairwise global alignment step that could lead to sub-optimal
alignments.
.TP
.BR v1.1\~ "released February 26, 2013"
Version 1.1 introduces two new important options: the possibility to
output swarming results using the uclust output format, and the
possibility to output detailed statistics on each swarms. Swarm 1.1 is
also faster: new filterings based on pairwise amplicon sequence
lengths and composition comparisons reduce the number of pairwise
alignments needed and speed up the swarming.
.TP
.BR v1.0\~ "released November 10, 2012"
First public release.
.LP
.\" ============================================================================
.\" NOTES
.\" visualize and output to pdf
.\" man -l swarm.1
.\" man -t ./swarm.1 | ps2pdf -sPAPERSIZE=a4 - > swarm_manual.pdf
.\"
.\" INSTALL (sysadmin)
.\" gzip -c swarm.1 > swarm.1.gz
.\" mv swarm.1.gz /usr/share/man/man1/


src/swarm.cc
/*
    SWARM

    Copyright (C) 2012-2015 Torbjorn Rognes and Frederic Mahe

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU Affero General Public License as
    published by the Free Software Foundation, either version 3 of the
    License, or (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU Affero General Public License for more details.

    You should have received a copy of the GNU Affero General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.

    Contact: Torbjorn Rognes <torognes@ifi.uio.no>,
    Department of Informatics, University of Oslo,
    PO Box 1080 Blindern, NO-0316 Oslo, Norway
*/

#include "swarm.h"

/* ARGUMENTS AND THEIR DEFAULTS */

#define DEFAULT_GAPOPEN 12
#define DEFAULT_GAPEXTEND 4
#define DEFAULT_MATCHSCORE 5
#define DEFAULT_MISMATCHSCORE (-4)
#define DEFAULT_THREADS 1
#define DEFAULT_RESOLUTION 1
#define DEFAULT_BREAK_SWARMS 0
#define DEFAULT_MOTHUR 0
#define DEFAULT_USEARCH_ABUNDANCE 0
#define DEFAULT_INTERNAL_STRUCTURE 0
#define DEFAULT_LOG 0
#define DEFAULT_NO_OTU_BREAKING 0
#define DEFAULT_FASTIDIOUS 0
#define DEFAULT_BOUNDARY 3

char * outfilename;
char * statsfilename;
char * uclustfilename;
char * progname;
char * databasename;
long gapopen;
long gapextend;
long matchscore;
long mismatchscore;
unsigned long threads;
long resolution;
long break_swarms;
long mothur;
long usearch_abundance;

char * opt_log;
char * opt_internal_structure;
long opt_no_otu_breaking;
long opt_fastidious;
long opt_boundary;

long penalty_factor;
long penalty_gapextend;
long penalty_gapopen;
long penalty_mismatch;

/* Other variables */

long mmx_present = 0;
long sse_present = 0;
long sse2_present = 0;
long sse3_present = 0;
long ssse3_present = 0;
long sse41_present = 0;
long sse42_present = 0;
long popcnt_present = 0;
long avx_present = 0;
long avx2_present = 0;

unsigned long dbsequencecount = 0;

FILE * outfile;
FILE * statsfile;
FILE * uclustfile;
FILE * logfile = stderr;
FILE * internal_structure_file;

char sym_nt[] = "-acgt                           ";

#define cpuid(f1, f2, a, b, c, d)                                       \
  __asm__ __volatile__ ("cpuid"                                         \
                        : "=a" (a), "=b" (b), "=c" (c), "=d" (d)        \
                        : "a" (f1), "c" (f2));

void cpu_features_detect()
{
  unsigned int a, b, c, d;

  cpuid(0, 0, a, b, c, d);
  unsigned int maxlevel = a & 0xff;

  if (maxlevel >= 1)
  {
    cpuid(1, 0, a, b, c, d);
    mmx_present    = (d >> 23) & 1;
    sse_present    = (d >> 25) & 1;
    sse2_present   = (d >> 26) & 1;
    sse3_present   = (c >>  0) & 1;
    ssse3_present  = (c >>  9) & 1;
    sse41_present  = (c >> 19) & 1;
    sse42_present  = (c >> 20) & 1;
    popcnt_present = (c >> 23) & 1;
    avx_present    = (c >> 28) & 1;

    if (maxlevel >= 7)
    {
      cpuid(7, 0, a, b, c, d);
      avx2_present   = (b >>  5) & 1;
    }
  }
}

void cpu_features_show()
{
  fprintf(logfile, "CPU features:     ");
  if (mmx_present)
    fprintf(logfile, " mmx");
  if (sse_present)
    fprintf(logfile, " sse");
  if (sse2_present)
    fprintf(logfile, " sse2");
  if (sse3_present)
    fprintf(logfile, " sse3");
  if (ssse3_present)
    fprintf(logfile, " ssse3");
  if (sse41_present)
    fprintf(logfile, " sse4.1");
  if (sse42_present)
    fprintf(logfile, " sse4.2");
  if (popcnt_present)
    fprintf(logfile, " popcnt");
  if (avx_present)
    fprintf(logfile, " avx");
  if (avx2_present)
    fprintf(logfile, " avx2");
  fprintf(logfile, "\n");
}


void args_getstring(int i, int argc, char **argv, char ** result, char * error)
{
  if (i+1 < argc)
    *result = argv[i+1];
  else
    fatal(error);
}

void args_getnum(int i, int argc, char **argv, long * result, char * error)
{
  if (i+1 < argc)
    *result = atol(argv[i+1]);
  else
    fatal(error);
}

void args_show()
{
  cpu_features_show();
  fprintf(logfile, "Database file:     %s\n", databasename ? databasename : "(stdin)");
  fprintf(logfile, "Output file:       %s\n", outfilename ? outfilename : "(stdout)");
  if (statsfilename)
    fprintf(logfile, "Statistics file:   %s\n", statsfilename);
  if (uclustfilename)
    fprintf(logfile, "Uclust file:       %s\n", uclustfilename);
  fprintf(logfile, "Resolution (d):    %ld\n", resolution);
  fprintf(logfile, "Threads:           %ld\n", threads);
  fprintf(logfile, "Scores:            match: %ld, mismatch: %ld\n", matchscore, mismatchscore);
  fprintf(logfile, "Gap penalties:     opening: %ld, extension: %ld\n", gapopen, gapextend);
  fprintf(logfile, "Converted costs:   mismatch: %ld, gap opening: %ld, gap extension: %ld\n", penalty_mismatch, penalty_gapopen, penalty_gapextend);
  fprintf(logfile, "Break OTUs:        %s\n", opt_no_otu_breaking ? "No" : "Yes");
  if (opt_fastidious)
    fprintf(logfile, "Fastidious:        Yes, with boundary %lu\n", opt_boundary);
  else
    fprintf(logfile, "Fastidious:        No\n");
}

void args_usage()
{
  /*               0         1         2         3         4         5         6         7          */
  /*               01234567890123456789012345678901234567890123456789012345678901234567890123456789 */

  fprintf(stderr, "Usage: %s [OPTIONS] [filename]\n", progname);
  fprintf(stderr, "  -d, --differences INTEGER           resolution (1)\n");
  fprintf(stderr, "  -h, --help                          display this help and exit\n");
  fprintf(stderr, "  -o, --output-file FILENAME          output result filename (stdout)\n");
  fprintf(stderr, "  -t, --threads INTEGER               number of threads to use (1)\n");
  fprintf(stderr, "  -v, --version                       display version information and exit\n");
  fprintf(stderr, "  -m, --match-reward INTEGER          reward for nucleotide match (5)\n");
  fprintf(stderr, "  -p, --mismatch-penalty INTEGER      penalty for nucleotide mismatch (4)\n");
  fprintf(stderr, "  -g, --gap-opening-penalty INTEGER   gap open penalty (12)\n");
  fprintf(stderr, "  -e, --gap-extension-penalty INTEGER gap extension penalty (4)\n");
  fprintf(stderr, "  -s, --statistics-file FILENAME      dump swarm statistics to file\n");
  fprintf(stderr, "  -u, --uclust-file FILENAME          output in UCLUST-like format to file\n");
  fprintf(stderr, "  -r, --mothur                        output in mothur list file format\n");
  fprintf(stderr, "  -z, --usearch_abundance             abundance annotation in usearch style\n");
  fprintf(stderr, "  -i, --internal-structure FILENAME   write internal swarm structure to file\n");
  fprintf(stderr, "  -l, --log FILENAME                  log to file, not to stderr\n");
  fprintf(stderr, "  -n, --no-otu-breaking               never break OTUs\n");
  fprintf(stderr, "  -f, --fastidious                    link nearby low-abundance swarms\n");
  fprintf(stderr, "  -b, --boundary INTEGER              min amplicon abundance for fastidious\n");
  fprintf(stderr, "\n");
  fprintf(stderr, "See 'man swarm' for more details.\n");
}

void show_header()
{
  char title[] = "Swarm " SWARM_VERSION;
  char ref[] = "Copyright (C) 2012-2015 Torbjorn Rognes and Frederic Mahe";
  char url[] = "https://github.com/torognes/swarm";
  fprintf(logfile, "%s [%s %s]\n%s\n%s\n\n",
          title, __DATE__, __TIME__, ref, url);
  fprintf(logfile, "Please cite: Mahe F, Rognes T, Quince C, de Vargas C, Dunthorn M (2014)\n");
  fprintf(logfile, "Swarm: robust and fast clustering method for amplicon-based studies.\n");
  fprintf(logfile, "PeerJ 2:e593 http://dx.doi.org/10.7717/peerj.593\n");
  fprintf(logfile, "\n");
}

void args_init(int argc, char **argv)
{
  /* Set defaults */

  progname = argv[0];

  databasename = NULL;
  outfilename = NULL;
  statsfilename = NULL;
  resolution = DEFAULT_RESOLUTION;
  threads = DEFAULT_THREADS;
  matchscore = DEFAULT_MATCHSCORE;
  mismatchscore = DEFAULT_MISMATCHSCORE;
  gapopen = DEFAULT_GAPOPEN;
  gapextend = DEFAULT_GAPEXTEND;
  mothur = DEFAULT_MOTHUR;
  usearch_abundance = DEFAULT_USEARCH_ABUNDANCE;
  opt_log = DEFAULT_LOG;
  opt_internal_structure = DEFAULT_INTERNAL_STRUCTURE;
  opt_no_otu_breaking = DEFAULT_NO_OTU_BREAKING;
  opt_fastidious = DEFAULT_FASTIDIOUS;
  opt_boundary = DEFAULT_BOUNDARY;

  opterr = 1;

  char short_options[] = "d:ho:t:vm:p:g:e:s:u:rzi:l:nfb:";

  static struct option long_options[] =
  {
    {"differences",           required_argument, NULL, 'd' },
    {"help",                  no_argument,       NULL, 'h' },
    {"output-file",           required_argument, NULL, 'o' },
    {"threads",               required_argument, NULL, 't' },
    {"version",               no_argument,       NULL, 'v' },
    {"match-reward",          required_argument, NULL, 'm' },
    {"mismatch-penalty",      required_argument, NULL, 'p' },
    {"gap-opening-penalty",   required_argument, NULL, 'g' },
    {"gap-extension-penalty", required_argument, NULL, 'e' },
    {"statistics-file",       required_argument, NULL, 's' },
    {"uclust-file",           required_argument, NULL, 'u' },
    {"mothur",                no_argument,       NULL, 'r' },
    {"usearch-abundance",     no_argument,       NULL, 'z' },
    {"internal-structure",    required_argument, NULL, 'i' },
    {"log",                   required_argument, NULL, 'l' },
    {"no-otu-breaking",       no_argument,       NULL, 'n' },
    {"fastidious",            no_argument,       NULL, 'f' },
    {"boundary",              required_argument, NULL, 'b' },
    { 0, 0, 0, 0 }
  };
  
  int option_index = 0;
  int c;
  
  while ((c = getopt_long(argc, argv, short_options, long_options, &option_index)) != -1)
  {
    switch(c)
    {
    case 'd':
      /* differences (resolution) */
      resolution = atol(optarg);
      break;
          
    case 'h':
      /* help */
      show_header();
      args_usage();
      exit(1);
      break;

    case 'o':
      /* output-file */
      outfilename = optarg;
      break;
          
    case 't':
      /* threads */
      threads = atol(optarg);
      break;
          
    case 'v':
      /* version */
      show_header();
      exit(1);
      break;

    case 'm':
      /* match reward */
      matchscore = atol(optarg);
      break;
          
    case 'p':
      /* mismatch penalty */
      mismatchscore = - atol(optarg);
      break;
          
    case 'g':
      /* gap opening penalty */
      gapopen = atol(optarg);
      break;
          
    case 'e':
      /* gap extension penalty */
      gapextend = atol(optarg);
      break;
          
    case 's':
      /* statistics-file */
      statsfilename = optarg;
      break;
          
    case 'u':
      /* uclust-file */
      uclustfilename = optarg;
      break;
          
    case 'r':
      /* mothur */
      mothur = 1;
      break;
          
    case 'z':
      /* usearch-abundance */
      usearch_abundance = 1;
      break;
          
    case 'i':
      /* internal-structure */
      opt_internal_structure = optarg;
      break;
          
    case 'l':
      /* log */
      opt_log = optarg;
      break;
          
    case 'n':
      /* no-otu-breaking */
      opt_no_otu_breaking = 1;
      break;
          
    case 'f':
      /* fastidious */
      opt_fastidious = 1;
      break;
          
    case 'b':
      /* boundary */
      opt_boundary = atol(optarg);
      break;
          
    default:
      show_header();
      args_usage();
      exit(1);
      break;
    }
  }
  
  if (optind < argc)
    databasename = argv[optind];
  
  if (resolution < 1)
    fatal("Illegal resolution specified.");

  if ((threads < 1) || (threads > MAX_THREADS))
    fatal("Illegal number of threads specified");
  
  if ((gapopen < 0) || (gapextend < 0) || ((gapopen + gapextend) < 1))
    fatal("Illegal gap penalties specified.");

  if (matchscore < 1)
    fatal("Illegal match reward specified.");

  if (mismatchscore > -1)
    fatal("Illegal mismatch penalty specified.");

  if (outfilename)
    {
      outfile = fopen(outfilename, "w");
      if (! outfile)
        fatal("Unable to open output file for writing.");
    }
  else
    outfile = stdout;
  
  if (statsfilename)
    {
      statsfile = fopen(statsfilename, "w");
      if (! statsfile)
        fatal("Unable to open statistics file for writing.");
    }
  else
    statsfile = 0;
  
  if (uclustfilename)
    {
      uclustfile = fopen(uclustfilename, "w");
      if (! uclustfile)
        fatal("Unable to open uclust file for writing.");
    }
  else
    uclustfile = 0;

  if (opt_log)
    {
      logfile = fopen(opt_log, "w");
      if (! logfile)
        fatal("Unable to open log file for writing.");
    }
  else
    logfile = stderr;

  if (opt_internal_structure)
    {
      internal_structure_file = fopen(opt_internal_structure, "w");
      if (! internal_structure_file)
        fatal("Unable to open internal structure file for writing.");
    }
  else
    internal_structure_file = stderr;

  if (opt_fastidious && (resolution != 1))
    fatal("The fastidious option only works when the resolution (d) is 1.\n");
}

int main(int argc, char** argv)
{
  cpu_features_detect();

  if (!sse2_present)
    fatal("This program requires a processor with SSE2 instructions.\n");
  
  args_init(argc, argv);

  penalty_mismatch = 2 * matchscore - 2 * mismatchscore;
  penalty_gapopen = 2 * gapopen;
  penalty_gapextend = 2 * matchscore + gapextend;

  penalty_factor = gcd(gcd(penalty_mismatch, penalty_gapopen), penalty_gapextend);
  
  penalty_mismatch /= penalty_factor;
  penalty_gapopen /= penalty_factor;
  penalty_gapextend /= penalty_factor;

  show_header();
  
  args_show();

  fprintf(logfile, "\n");

  db_read(databasename);
  
  fprintf(logfile, "Database info:     %ld nt", db_getnucleotidecount());
  fprintf(logfile, " in %ld sequences,", db_getsequencecount());
  fprintf(logfile, " longest %ld nt\n", db_getlongestsequence());

  dbsequencecount = db_getsequencecount();

  score_matrix_init();

  search_begin();
  
  if (resolution == 1)
    algo_d1_run();
  else
    algo_run();

  search_end();

  score_matrix_free();

  db_free();

  if (uclustfile)
    fclose(uclustfile);

  if (statsfile)
    fclose(statsfile);

  if (outfilename)
    fclose(outfile);
}
